Messages Cluster XML

lazarkov opened the pull request:
This is continuation of PR https://github.com/project-chip/connectedhomeip/pull/30254/

It's all about defining Messages cluster XML and creating boilerplate code using ZAP.

andy31415 commented on src/app/clusters/messages-server/messages-server.h:
```diff
@@ -0,0 +1,29 @@
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Messaging {} // namespace Messaging
```
is there any use having these empty namespaces here?

lazarkov replied:
it's a copy pasta from https://github.com/project-chip/connectedhomeip/pull/30254/. However as I plan to enable this cluster similarly to the previous clusters, this name space will be extended with some methods.

andy31415 commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,95 @@ provisional cluster DemandResponseLoadControl = 150 {
 
+/** This cluster provides an interface for passing messages to be presented by a device. */
+provisional cluster Messages = 151 {
+  revision 1;
```
Messages is at revision 3 in the spec.

This results in several diffs I believe.

lazarkov replied:
Updated.

andy31415 commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,95 @@ provisional cluster DemandResponseLoadControl = 150 {
+  }
+
+  bitmap Feature : bitmap32 {
+    kReceivedConfirmation = 0x1;
```
all bitmaps seem wrong: they should be masks shifted, not enum-like

lazarkov replied:
Updated.

andy31415 reviewed (changes requested):
Please validate the XML for the spec. Looking at the generated .matter (so that it is human readable) it generally looks wrong.

andy31415 commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,95 @@ provisional cluster DemandResponseLoadControl = 150 {
+  /** Command for requesting messages be presented */
+  command PresentMessagesRequest(PresentMessagesRequestRequest): DefaultSuccess = 0;
+  /** Command for cancelling message present requests */
+  command CancelMessagesRequest(CancelMessagesRequestRequest): DefaultSuccess = 1;
```
commands should be flagged as fabric

lazarkov replied:
Updated.

lazarkov commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,95 @@ provisional cluster DemandResponseLoadControl = 150 {
+  readonly attribute int16u clusterRevision = 65533;
+
+  request struct PresentMessagesRequestRequest {
+    octet_string messages[] = 0;
```
Yup, I've forgot to save this edit locally. Updated.

andy31415 replied:
This is not resolved

lazarkov replied:
Hm, updated. now

andy31415 replied:
Still not updated....

lazarkov replied:
Checked that:
- attribute ActiveMessageIDs should be octect_string and they are
- CancelMessagesRequest MessageIDs should be octect_string and they are
- MessageQueued event MessageID should be octect_string and it is
- MessagePresented event MessageID should be octect_string and it is
- MessageComplete event MessageID should be octect_string and it is
- PresentMessagesRequest has messages that are of type MessageStruct

lazarkov commented:
@andy31415 
> Please validate the XML for the spec. Looking at the generated .matter (so that it is human readable) it generally looks wrong. I added some comments, however not sure how much is because "revision 1 vs 3" or because missing bits in XML.
> 
> I expect we want latest revision.

I've run the suggested code and updated the XML accordingly. 

`./scripts/py_matter_idl/matter_idl/data_model_xml_parser.py \
  -o out/spec.matter \
  --compare-output out/sdk.matter \
  --compare src/controller/data_model/controller-clusters.matter \
  data_model/clusters/Messages.xml

diff out/{spec,sdk}.matter`

**Note that spec still has events, but auto generated Messages.xml does not. However I kept the events in messages-cluster.xml.**

andy31415 commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,95 @@ provisional cluster DemandResponseLoadControl = 150 {
+  info event MessageComplete = 2 {
+    octet_string messageID = 0;
+    epoch_s timestamp = 2;
+    nullable int32u responseID = 3;
```
ids seem off - my copy of the spec has responseID as 1, there is no timestamp and several other changes.

lazarkov replied:
Yes, updated.

lazarkov commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,95 @@ provisional cluster DemandResponseLoadControl = 150 {
+  }
+
+  info event MessageQueued = 0 {
+    octet_string messageID = 0;
```
Updated.

andy31415 replied:
Does not look like the changes were made...

lazarkov replied:
I've tried with max and with length argument both of which do not work.
`<field id="0" name="MessageID" type="octet_string" max="16"/>`
`<field id="0" name="MessageID" type="octet_string" length="16"/>`

andy31415 replied:
thermostat-cluster.xml has a `<item fieldId="0" name="ScheduleHandle" type="octet_string" length="16" isNullable="true"/>` and it works, so it seems like "length" should work (although I am not sure how field,item and arg interact and what supports what...

lazarkov replied:
I haven't found single event that has tag item, they all have tag field. 

I'm usually using test-cluster.xml for checking what could be done.

lazarkov commented:
I've created for ZAP project isFabricScoped issue:
https://github.com/project-chip/zap/issues/1255

bzbarsky-apple commented:
@lazarkov Why is this a new PR instead of continuing with the existing PR that had existing review comments?

lazarkov commented on src/app/clusters/messages-server/messages-server.cpp:
Yes. It came from original PR, however it won't harm as I'll plan use it in next PR

michaelleecowan reviewed (approved):
approve

lazarkov commented on src/app/zap-templates/zcl/data-model/chip/messages-cluster.xml:
```diff
@@ -0,0 +1,111 @@
+    <client tick="false" init="false">true</client>
+    <server tick="false" init="false">true</server>
+    <globalAttribute side="either" code="0xFFFD" value="3"/><!-- Revision   -->
+    <attribute side="server" code="0x0000" define="LIST_MESSAGES" type="array" entryType="MessageStruct" length="8" writable="true" optional="false">Messages</attribute>
```
I've used script that Andrei suggested. However the diff it creates is not correct (somehow events were removed) or it suggests to add attributes that are not supported.

lazarkov replied:
Also updated the attribute writable to false.

andy31415 commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,93 @@ provisional cluster DemandResponseLoadControl = 150 {
+  info event MessageComplete = 2 {
+    octet_string messageID = 0;
+    optional nullable int32u responseID = 1;
+    optional nullable char_string reply = 2;
```
since char_string is max 254 in ember due to pascal-strings, this probably should be long_char_string to be able to hold the entire spec max of 256

andy31415 replied:
This may be ok here and for the struct since it is not stored by ember in its attribute storage. Marking resolved, however unsure if we should mark this as long_char_string or not. Is this one of the cases when length cannot be set?

andy31415 commented on src/controller/data_model/controller-clusters.matter:
```diff
@@ -4372,6 +4372,93 @@ provisional cluster DemandResponseLoadControl = 150 {
+    optional char_string<32> label = 1;
+  }
+
+  struct MessageStruct {
```
this struct is fabric-scoped in the specification. Is there a specification issue to turn this in non-fabric scoped due to issues including in command or what is the plan here?

Otherwise this must be flagged as such (and we need a fabric_idx as field 254)

chrisdecenzo replied:
see https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/8806

andy31415 commented on src/controller/java/zap-generated/CHIPInvokeCallbacks.cpp:
This file was removed in #31675 ... we should probably make efforts to detect re-additions.
